storagebackend/factory: Simplify etcd health check

tchap opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The function is unnecessarily complex just because the gRPC client is
initialized with `grpc.WithBlock()`, which is unnecessary in this case
since there is anyway a health check request sent to etcd, which times
out in case the connection is not established yet.

#### Which issue(s) this PR is related to:

N/A

#### Special notes for your reviewer:

N/A

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

N/A

tchap committed:
storagebackend/factory: Simplify etcd health check

The function is unnecessarily complex just because the gRPC client is
initialized with grpc.WithBlock(), which is unnecessary in this case
since there is anyway a health check request sent to etcd, which times
out in case the connection is not established yet.

aaron-prindle commented:
/sig etcd

wojtek-t commented:
/assign @serathius

serathius commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -151,69 +152,48 @@ func (a *atomicLastError) Load() error {
+	// Create a new prober monitor.
+	prober, err := newETCD3ProberMonitor(c)
+	if err != nil {
+		return nil, err
```
Now the function might exit early, so I double checked when `newETCD3ProberMonitor` returns error. There are 2 potential sources of errors: egressLookup and client constructor. Both look fine.

serathius commented:
/lgtm

serathius commented:
From pull-kubernetes-unit
```
=== RUN   TestKMSHealthzEndpoint/no_kms-provider,_expect_no_kms_healthz_check,_no_kms_livez_check
    etcd_test.go:366: Failed to add healthz error: etcdclient: no available endpoints
```
As I mentioned above your change adds option to exit early if client creation errors. I check and errors should be just connected to invalid configuration, but it seems we have multiple tests that didn't configure client correctly.

To make the PR pass tests you need to fix the etcd client config in other tests.

tchap commented:
> From pull-kubernetes-unit
> 
> ```
> === RUN   TestKMSHealthzEndpoint/no_kms-provider,_expect_no_kms_healthz_check,_no_kms_livez_check
>     etcd_test.go:366: Failed to add healthz error: etcdclient: no available endpoints
> ```

This is now fixed. I added non-existent endpoints, which is ok for these tests it seems.

tchap commented:
> From [pull-kubernetes-linter-hints](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/133709/pull-kubernetes-linter-hints/1963215883719413760)
> 
> ```
> ERROR: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:172:15: Error return value of `prober.Close` is not checked (errcheck)
> ERROR: 		prober.Close()
> ERROR: 		            ^
> ERROR: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:367:36: SA1019: grpc.WithBlock is deprecated: this DialOption is not supported by NewClient. Will be supported throughout 1.x. (staticcheck)
> ERROR: 		client, err := newETCD3Client(c, grpc.WithBlock())
> ERROR: 		                                 ^
> ERROR: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:411:45: SA1019: grpc.WithBlock is deprecated: this DialOption is not supported by NewClient. Will be supported throughout 1.x. (staticcheck)
> ERROR: 	client, err := newETCD3Client(c.Transport, grpc.WithBlock())
> ```

Yeah, I know, but I am not sure much can be done about it. The problem was already there regarding `WithBlock`, it's just now triggered in CI because the relevant lines have changed. You I will see whether anything can be done about it. This is not a required CI job, though...

tchap commented:
I fixed the `probe.Close()` linter error, but I honestly think that fixing `WithBlock` is out of scope for this PR, because it changes behavior of quite a few functions that suddenly stop failing on dial when the connection cannot be established.

tchap commented:
/retest-required

serathius commented:
> I fixed the probe.Close() linter error, but I honestly think that fixing WithBlock is out of scope for this PR, because it changes behavior of quite a few functions that suddenly stop failing on dial when the connection cannot be established.

You can disable linter on that line.

tchap commented:
@serathius All CI jobs are passing now, disabled linting where necessary.

serathius commented:
/lgtm

/cc @liggitt @wojtek-t

tchap commented:
/retest-required

tchap commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -283,7 +263,7 @@ func (t *etcd3ProberMonitor) Monitor(ctx context.Context) (metrics.StorageMetric
 }
 
-var newETCD3Client = func(c storagebackend.TransportConfig) (*kubernetes.Client, error) {
+var newETCD3Client = func(c storagebackend.TransportConfig, grpcOpts ...grpc.DialOption) (*kubernetes.Client, error) {
```
Amended.

Jefftree commented:
/cc @siyuanfoundation 
/triage accepted

tchap commented:
This is starting to rot, could somebody please take a look for review? Thanks.
